### ISSUE TITLE
Rewrite PyTorch CUDA backend in Triton

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#62661]
@chsasank opened this issue in https://github.com/openai/triton/issues/150 and I thought this was pretty interesting. It's true that to actually ship this in PyTorch mainline, we would indeed want an offline Triton compiler, and we'd also need to work out all of the distribution shenanigans that a project as big as Python would have to do.

However, I think it is feasible to experimentally prototype Triton versions of PyTorch kernels out of tree, with only a few extra hooks in PyTorch core. In https://github.com/pytorch/pytorch/pull/62660 I give a prototype for how to directly register Python implementations of kernels, allowing us to *override* preexisting CUDA implementations. So here's how it would work:

1. You write a Triton kernel to replace some kernel in PyTorch
2. You call the Python operator registration API, and override the existing CUDA kernel with your new Triton kernel
3. Profit!

There's a little example of how to override existing kernels in https://github.com/pytorch/pytorch/pull/62660/files#diff-415017bcad4fa6cd6d3dfe5f6ea1caffcd7122b46b8c1e4825f7d889efc80a62 (the API needs improvement). To make it work with Triton, you would just call into Triton in the same way as the tutorial. I could probably produce a working end to end example, but I'll let whatever intrepid soul who wants to embark on this project have the fun of figuring that out.

If someone is actually interested in setting up the repo to do this and charging ahead, I'll prioritize getting #62660 into PyTorch proper.

cc @ptillet

**Comment by chsasank** [comment]
What'd be a good benchmark/PoC that we can build with the triton kernel based pytorch? I think it'll help to have an achievable goal that can validate the idea in the first place.

**Comment by ptillet** [comment]
An AOT module has been requested a few times already. I've been a bit busy writing kernels recently, but I'm sure I could find some time to help with that. What I have in mind would be an API similar to https://numba.pydata.org/numba-doc/dev/user/pycc.html which would work as follows:

my-kernels.py
```
@triton.aot('f32*,f32', META={'BLOCK': 1024}) # register for AOT compilation for float32 signature with block size 1024.
@triton.jit # makes it callable just-in-time
def _kernel1(x, y, **META):
  ...
 
# Do the same as above -- will be packaged in the same shared library
@triton.aot('f16*')
@triton.jit
def _kernel2(z, **META):
  ...
  
```

And then one would do something like
`tcc my-kernels.py --arch=sm_80 --api=c -o my-kernels.so --headers my-kernels.h`

and the output would be a `.so` shared libraries which defines the functions declared in the header:

```
void _kernel1_BLOCK1024(tl_float32*, tl_float32);

void _kernel2(tl_float16*, tl_float16);
```

And that would also serve as a C API to Triton, as a bunch of people have complained already that Triton isn't usable in C/C++ codebase.

We could also have something like
```
`tcc my-kernels.py --arch=sm_80 --api=python -o my-kernels.so`
```
to generate some python bindings automatically (maybe via pybind11) but this seems somewhat lower priority.

Does that kind of API sound good and sufficient?

**Comment by ezyang** [comment]
> What'd be a good benchmark/PoC that we can build with the triton kernel based pytorch? I think it'll help to have an achievable goal that can validate the idea in the first place.

* As a start, I'd take the tutorial kernels from Triton which have already been benchmarked to be better implemented in Triton than in PyTorch and drop 'em in and show end to end performance increase (this might tricky with Python based registration, more on this shortly).
* Try looking at the really crappy CUDA kernels that were written just because we needed *something* that wouldn't cause a DtH synchronization and look for 100x or 1000x speedups there. Look for stuff that was written recently, or just put your CUDA expert hat on and look for kernels that just don't look very good.
* Check out the intersection of performance and cuda issues is a good place to start: https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+performance%22+label%3A%22module%3A+cuda%22 You could also just pick some of your favorite models and see if you can make them go faster.
* Reimplementations of cuDNN/cuBLAS/etc kernels is generally useful, because these are often the types of kernels people want to customize and cannot do so because they're locked up in vendor libraries. For example, we still don't have a lot of variations of matrix multiply e.g. at https://github.com/pytorch/pytorch/issues/29961 and https://github.com/pytorch/pytorch/issues/61909

> Does that kind of API sound good and sufficient?

Yes, this seems good. Integrating directly with the C API will be better than calling into the Python JIT because we get to skip the overhead of doing error checking etc in Python and let us get a really realistic understanding of how much faster Triton is (at the cost of this Triton-PyTorch extension requiring compilation to distribute!)

**Comment by chsasank** [comment]
I'm thinking something like triton-kernel only imagenet training code. But that'll be a tall order 😅. And it does require us to write some cudnn/cublas kernels in triton.

**Comment by ptillet** [comment]
@chsasank There used to be some conv2d/conv3d Triton implementation that was on par with cuDNN, but it got scrapped off because of the maintenance burden it was. I think that if you want some kind of end-to-end triton-only model, transformer would be the right choice. All the building blocks should already pretty much already be there.

**Comment by chsasank** [comment]
> There used to be some conv2d/conv3d Triton implementation that was on par with cuDNN, but it got scrapped off because of the maintenance burden it was.

Well if we're successful, we can write triton-blas and triton-nn libraries :) 

> transformer would be the right choice

Yep transformers make lot of sense. All we really need are matrix multiplications, softmax and relu.

**Comment by ezyang** [comment]
> Yep transformers make lot of sense. All we really need are matrix multiplications, softmax and relu. We still need to write grad functions though right?

Yes. Fortunately in the case of transformer, they're pretty simple, and you don't need AD, PyTorch can take care of that for you.

**Comment by ptillet** [comment]
> Yep transformers make lot of sense. All we really need are matrix multiplications, softmax and relu. We still need to write grad functions though right?

I think you'd still need to write GPU kernels for backward pass. But the Triton samples already have this (except the softmax I believe -- although we have softmax + cross-entropy and block-sparse softmax)

Feel free to join the dev slack at https://join.slack.com/t/triton-lang/shared_invite/zt-tuoocpwz-kojxc~Lw6L6oZBtZ7wIiJw if you want. I'm not there all the time but it'll be a more direct channel for any specific issue you face while playing with Triton.

**Comment by vadimkantorov** [comment]
One good example of custom needed kernels could be https://github.com/pytorch/pytorch/issues/55366 or matmul/cdist-like ops with custom elementwise, reduction and transform ops (e.g. elementwise min + sum) + this could be reused for sth like pykeops

**Comment by ngimel** [comment]
>Check out the intersection of performance and cuda issues is a good place to start: https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+performance%22+label%3A%22module%3A+cuda%22 You could also just pick some of your favorite models and see if you can make them go faster.

I've spot-checked a few of these issues, and didn't find good candidates, they are mostly "D2H copy is slow" and "Ampere is slow". The latter class of issues is potentially interesting (the root cause was torch dropping some kernels from cudnn with static linking, and cudnn missing fast kernels in earlier versions, both are not fixed in nightlies that uses 8.2), but would require writing convolutions which is probably not an ideal first step. 
Matmul-like operations with custom reductions are also interesting, but currently afaiu `.dot` in triton does only pointwise multiply + add,  and changing that to do e.g. max-sum (#59545) is also non-trivial.

**Comment by vadimkantorov** [comment]
@ezyang @ngimel maybe then we could have a separate contrib repo under PyTorch organization with both Numba and Triton collection of kernels with tested utils for interop/dispatch and examples (scatters / generalized matmul, dtype specialization)? this would allow for a simpler experimentation for users (and no problems of static linking and bloated PyTorch binary size) and would allow Numba and Triton devs gain feedback on more usecases

**Comment by chsasank** [comment]
I've started to write some functions based on triton's tutorials here: https://github.com/chsasank/triton-torch. My ops doesn't support autograd yet. [Is this the recommended way](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd) to get autograd support for my ops? I suppose I'll have to write kernels for backward of softmax. Any code reference in pytorch code base I can use for this?

**Comment by ezyang** [comment]
Nice stuff. With the shallow integration that you're doing in your repository, unfortunately the only way to get autograd going is define custom functions (as you see in the docs you linked.) So for example:

```
class Add(torch.autograd.Function):
  @staticmethod
  def forward(ctx, x: torch.Tensor, y: torch.Tensor):
    output = torch.empty_like(x)
    assert x.is_cuda and y.is_cuda and output.is_cuda
    n_elements = output.numel()
    grid = lambda meta: (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
    add_kernel[grid](x, y, output, n_elements, BLOCK_SIZE=1024)
    return output

  @staticmethod
  def backward(ctx, grad_output):
    return (grad_output, grad_output)
```

With the deeper integration (actually overriding the behavior of torch.add) we can hook into the preexisting autograd functions defined in PyTorch, but torch.autograd.Function is a good way to get started and we can always refactor later.

**Comment by chsasank** [comment]
Thanks. I'm intimidated by pytorch code base. Can you please help me find my way around for softmax backward kernel? I have gone through [this](http://blog.ezyang.com/2019/05/pytorch-internals/) but still confused :(.

**Comment by ezyang** [comment]
You're going to have to write the softmax backward kernel, unfortunately. The preexisting softmax backward CUDA implementation lives at aten/src/ATen/native/cuda/SoftMax.cu but it's pretty involved...  Maybe https://eli.thegreenplace.net/2016/the-softmax-function-and-its-derivative/ is helpful

**Comment by ptillet** [comment]
@chsasank softmax backward is pretty easy to write in Triton. see https://github.com/openai/triton/blob/master/python/triton/ops/blocksparse/softmax.py#L109-L114 specifically

**Comment by ngimel** [comment]
@ptillet would the softmax as written in those examples work for large softmax axis, when values can no longer be held in registers/shared memory? If not, how would one go about picking between that implementation, and less efficient one where reductions are done separately, followed by pointwise transformation?

**Comment by ptillet** [comment]
@ngimel You are right, it wouldn't work. Using too much registers is less of an issue since the program will still run -- albeit slower -- since the compiler can just spill. Using too much shared memory will result in an error though. There are two possible solutions:

- Check whether `num_cols * bytes_per_elt < max_shared_mem`. Works but not great since it explicitly depends on details (e.g., amount of shared memory used by the kernel) that should be completely hidden to the user
- Catch [triton.OutOfResources](https://github.com/openai/triton/blob/f26a48a3b408b219dbf3e187ef14b7f71fd59de5/python/triton/code_gen.py#L457) when the kernel is executed for the first time, set a flag and fallback to the more general implementations on all subsequent calls. This is a better solution, but note that it still requires compiling a kernel that may not be used.

**Comment by ptillet** [comment]
Edit: probably better to not use a flag in the second case, and just catch the exception all the time.

**Comment by vadimkantorov** [comment]
A useful one: fused matmul + argmax, useful for simple retrieval

**Comment by chsasank** [comment]
I guess we can close this now that pytorch2.0 has triton backend :)

**Comment by ezyang** [comment]
Still got plenty of boutique kernels we gotta write Triton templates for :)
